server/src/main/java/org/elasticsearch/common/time/DateUtils.java

```diff
@@ -293,6 +293,17 @@ public static long toMilliSeconds(long nanoSecondsSinceEpoch) {
         return nanoSecondsSinceEpoch / 1_000_000;
     }
 
+    public static int compareNanosToMillis(long nanos, long millis) {
```

Comment:
This is the main implementation.  All the specific operators are implemented in terms of this method.